source/zipkin/zipkin_core_types.cc

```diff
@@ -0,0 +1,255 @@
+}
+
+Span::Span(const Span& span) {
+  trace_id_ = span.traceId();
+  name_ = span.name();
+  id_ = span.id();
+  parent_id_ = span.parentId();
+  annotations_ = span.annotations();
+  binary_annotations_ = span.binaryAnnotations();
+  debug_ = span.isDebug();
+  timestamp_ = span.timestamp();
+  duration_ = span.duration();
+  trace_id_high_ = span.traceIdHigh();
+  isset_ = span.isSet();
+  start_time_ = span.startTime();
+  tracer_ = span.tracer();
+}
+
+Span& Span::operator=(const Span& span) {
+  trace_id_ = span.traceId();
+  name_ = span.name();
+  id_ = span.id();
+  parent_id_ = span.parentId();
+  annotations_ = span.annotations();
+  binary_annotations_ = span.binaryAnnotations();
+  debug_ = span.isDebug();
+  timestamp_ = span.timestamp();
+  duration_ = span.duration();
+  trace_id_high_ = span.traceIdHigh();
+  isset_ = span.isSet();
+  start_time_ = span.startTime();
+  tracer_ = span.tracer();
+
+  return *this;
+}
+
+const std::string& Span::toJson() {
+  rapidjson::StringBuffer s;
+  rapidjson::Writer<rapidjson::StringBuffer> writer(s);
+  writer.StartObject();
+  writer.Key(ZipkinJsonFieldNames::SPAN_TRACE_ID.c_str());
+  writer.String(Util::uint64ToBase16(trace_id_).c_str());
+  writer.Key(ZipkinJsonFieldNames::SPAN_NAME.c_str());
+  writer.String(name_.c_str());
+  writer.Key(ZipkinJsonFieldNames::SPAN_ID.c_str());
+  writer.String(Util::uint64ToBase16(id_).c_str());
+
+  if (isset_.parent_id && parent_id_) {
+    writer.Key(ZipkinJsonFieldNames::SPAN_PARENT_ID.c_str());
+    writer.String(Util::uint64ToBase16(parent_id_).c_str());
+  }
+
+  if (isset_.timestamp) {
+    writer.Key(ZipkinJsonFieldNames::SPAN_TIMESTAMP.c_str());
+    writer.Int64(timestamp_);
+  }
+
+  if (isset_.duration) {
+    writer.Key(ZipkinJsonFieldNames::SPAN_DURATION.c_str());
+    writer.Int64(duration_);
+  }
+
+  if (isset_.debug) {
+    writer.Key(ZipkinJsonFieldNames::SPAN_DEBUG.c_str());
+    writer.Bool(debug_);
+  }
+
+  if (isset_.trace_id_high) {
+    writer.Key(ZipkinJsonFieldNames::SPAN_TRACE_ID_HIGH.c_str());
+    writer.Int64(trace_id_high_);
+  }
+
+  writer.EndObject();
+
+  json_string_ = s.GetString();
+
+  std::cerr << "Will convert annotations to JSON" << std::endl;
+
+  std::vector<const std::string*> annotation_json_vector;
+
+  for (auto it = annotations_.begin(); it != annotations_.end(); it++) {
+    annotation_json_vector.push_back(&(it->toJson()));
+  }
+  Util::addArrayToJson(json_string_, annotation_json_vector,
+                       ZipkinJsonFieldNames::SPAN_ANNOTATIONS.c_str());
+
+  std::vector<const std::string*> binary_annotation_json_vector;
+  for (auto it = binary_annotations_.begin(); it != binary_annotations_.end(); it++) {
+    binary_annotation_json_vector.push_back(&(it->toJson()));
+  }
+  Util::addArrayToJson(json_string_, binary_annotation_json_vector,
+                       ZipkinJsonFieldNames::SPAN_BINARY_ANNOTATIONS.c_str());
+
+  return json_string_;
+}
+
+void Span::finish() {
```

Comment:
when an operation abends, is finish still called? if so, we'd usually put a string binary annotation of key "error" and value of the error message in the span